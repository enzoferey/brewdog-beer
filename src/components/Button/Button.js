import React from "react";
import PropTypes from "prop-types";
import style from "./button.scss";

class Button extends React.Component {
  state = {
    countdown: 0,
    paused: false,
    done: false,
  };

  componentWillUnmount = () => {
    this.clearCountdown();
  };

  onClick = () => {
    const { countdown, paused, done } = this.state;
    const { delay, waiting } = this.props;

    if (!waiting) {
      if (delay > 0 && !done) {
        if (countdown === 0) {
          // first click, start
          this.setCountdown(delay);
        } else if (!paused) {
          // second click, pause
          this.setState({ paused: true });
          this.clearCountdown();
        } else {
          // third click, unpause
          this.setState({ paused: false });
          this.setCountdown(countdown);
        }
      } else {
        this.setState({ done: true });
      }
    }
  };

  clearCountdown = () => {
    clearInterval(this._countdown);
  };

  setCountdown = countdown => {
    this.setState({ countdown }, () => {
      this._countdown = setInterval(this.updateCountdown, 1000);
    });
  };

  updateCountdown = () => {
    const { countdown } = this.state;

    // Update value
    const newCountdown = countdown - 1;
    this.setState({ countdown: newCountdown });

    // Check donw
    if (newCountdown === 0) {
      this.setState({ done: true });
      this.clearCountdown();
    }
  };

  getLabel = () => {
    const { done, countdown } = this.state;
    const { waiting } = this.props;

    if (waiting) return "Complete prev";
    if (done) return "Added !";
    if (countdown > 0) return `${countdown} s`;
    return "Add";
  };

  getWidth = () => {
    const { countdown } = this.state;
    const { delay } = this.props;

    return 100 - (countdown / delay) * 100;
  };

  render = () => {
    const { countdown } = this.state;
    const { waiting } = this.props;

    return (
      <div
        className={`${style.main} ${waiting ? style.waiting : ""}`}
        onClick={this.onClick}
      >
        <p className={style.label}>{this.getLabel()}</p>
        <div
          style={{
            width: `${this.getWidth()}%`,
          }}
          className={`${style.progress} ${countdown > 0 ? style.active : ""}`}
        />
      </div>
    );
  };
}

Button.propTypes = {
  delay: PropTypes.number,
  waiting: PropTypes.bool,
};

export default Button;
