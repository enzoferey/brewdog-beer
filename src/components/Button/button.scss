.main {
  display: inline-block;
  position: relative;
  width: 170px;
  height: 50px;
  background-color: #f44336;
  cursor: pointer;
  border-radius: 4px;
  overflow: hidden;

  &.waiting {
    background-color: grey;
    cursor: not-allowed;
  }
}

.label {
  position: absolute;
  left: 50%;
  top: 50%;
  transform: translate(-50%, -50%);
  margin: 0;
  z-index: 2;
  color: white;
  font-weight: bold;
  font-size: 20px;
  text-align: center;
  width: 100%;
}

.progress {
  position: absolute;
  left: 0;
  top: 0;
  height: 100%;
  z-index: 1;
  background: black;
  opacity: 0;
  transition: opacity 0.3s ease;

  &.active {
    opacity: 0.1;
  }
}
