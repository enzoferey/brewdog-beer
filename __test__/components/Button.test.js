import React from "react";
import Button from "components/Button";

let done = false;
const callback = jest.fn(() => {
  done = true;
});

const props = {
  done,
  delay: 0,
  waiting: false,
  callback,
  background: "pink",
};

// Fake setInterval
jest.useFakeTimers();

test("renders correctly", () => {
  testSnapshot(<Button {...props} />);
});

describe("click interation", () => {
  describe("when click for the firt time", () => {
    describe("if `delay` prop === 0", () => {
      const wrapper = shallow(<Button {...props} />);

      it("should just call the callback", () => {
        wrapper.simulate("click");
        expect(callback).toHaveBeenCalledTimes(1);
      });
    });

    describe("if `delay` prop > 0", () => {
      const delay = 10;
      const wrapper = shallow(<Button {...props} delay={delay} />);

      it("start the timer", () => {
        wrapper.simulate("click");
        expect(setInterval).toHaveBeenCalledTimes(1);
        expect(wrapper.state("countdown")).toBe(delay);

        jest.clearAllTimers();
      });
    });
  });
});

describe("countdown", () => {
  it("should decrease countdown by one every second", () => {
    const delay = 10;
    const wrapper = shallow(<Button {...props} delay={delay} />);

    // Activate countdown
    wrapper.simulate("click");

    // Fast-forward until next setInterval tick
    jest.runTimersToTime(1000);

    expect(wrapper.state("countdown")).toBe(delay - 1);
  });

  it("should call the callback prop when equals 0", () => {
    // Reset call times
    callback.mockClear();

    const delay = 3;
    const wrapper = shallow(<Button {...props} delay={delay} />);

    // Activate countdown
    wrapper.simulate("click");

    // Fast-forward until countdown is 0
    jest.runTimersToTime(delay * 1000);

    expect(wrapper.state("countdown")).toBe(0);
    expect(callback).toHaveBeenCalledTimes(1);
  });
});
