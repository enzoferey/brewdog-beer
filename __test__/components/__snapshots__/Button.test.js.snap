// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="undefined "
  onClick={[Function]}
  style={
    Object {
      "background": "pink",
    }
  }
>
  <p>
    Add
  </p>
  <div
    className="undefined "
    style={
      Object {
        "width": "NaN%",
      }
    }
  />
</div>
`;
